src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -34,41 +36,170 @@ public static class TagHelperDescriptorFactory
         {
-            var tagNames = GetTagNames(type);
-            var typeName = type.FullName;
+            var typeInfo = type.GetTypeInfo();
             var attributeDescriptors = GetAttributeDescriptors(type);
+            var targetElementAttributes = GetValidTargetElementAttributes(typeInfo, errorSink);
+            var tagHelperDescriptors =
+                BuildTagHelperDescriptors(
+                    typeInfo,
+                    assemblyName,
+                    attributeDescriptors,
+                    targetElementAttributes);
+
+            return tagHelperDescriptors.Distinct(TagHelperDescriptorComparer.Default);
+        }
+
+        public static ICollection<char> InvalidNonWhitespaceNameCharacters { get; } = new HashSet<char>(
+            new[] { '@', '!', '<', '/', '?', '[', '>', ']', '=', '"', '\'' });
+
+        private static IEnumerable<TargetElementAttribute> GetValidTargetElementAttributes(
+            TypeInfo typeInfo,
+            ParserErrorSink errorSink)
+        {
+            var targetElementAttributes = typeInfo.GetCustomAttributes<TargetElementAttribute>(inherit: false);
+
+            return targetElementAttributes.Where(attribute => ValidTargetElementAttributeNames(attribute, errorSink));
+        }
+
+        private static IEnumerable<TagHelperDescriptor> BuildTagHelperDescriptors(
+            TypeInfo typeInfo,
+            string assemblyName,
+            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
+            IEnumerable<TargetElementAttribute> targetElementAttributes)
+        {
+            var typeName = typeInfo.FullName;
+
+            // If there isn't an attribute specifying the tag name derive it from the name
+            if (!targetElementAttributes.Any())
+            {
+                var name = typeInfo.Name;
+
+                if (name.EndsWith(TagHelperNameEnding, StringComparison.OrdinalIgnoreCase))
+                {
+                    name = name.Substring(0, name.Length - TagHelperNameEnding.Length);
+                }
+
+                return BuildTagHelperDescriptors(
+                    tagNames: new[] { ToHtmlCase(name) },
+                    typeName: typeName,
+                    assemblyName: assemblyName,
+                    attributeDescriptors: attributeDescriptors,
+                    requiredAttributes: Enumerable.Empty<string>());
+            }
+
+            return targetElementAttributes.SelectMany(
+                attribute => BuildTagHelperDescriptors(typeName, assemblyName, attributeDescriptors, attribute));
+        }
+
+        private static IEnumerable<TagHelperDescriptor> BuildTagHelperDescriptors(
+            string typeName,
+            string assemblyName,
+            IEnumerable<TagHelperAttributeDescriptor> attributeDescriptors,
+            TargetElementAttribute targetElementAttribute)
+        {
+            var tagNames = GetCommaSeparatedValues(targetElementAttribute.Tag);
```

Comment:
`Tag` is no longer a comma-separated value.  so this method and the `BuildTagHelperDescriptors` overload it calls can now return a single `TagHelperDescriptor`.  line 95 above can also change from `SelectMany` to `Select`.